### DISCUSSION TITLE
cammini

### DISCUSSION THREAD

**Discussion by IonutCicio** [sapienzastudentsnetwork/progettazione-di-algoritmi#5] [Esercizi A-L]
Dato un grafo $G = (V, E)$ e due vertici $x, y \in V$ scrivere lo pseudocodice di un algoritmo che determina se esiste un cammino da $x$ a $y$ in $O(|V| + |E|)$

**Comment by IonutCicio** [comment]
Dato un grafo $G$ come lista di adiacenza, e due vertici $x$, $y$ nel grafo, l'algoritmo fa una visita partendo da $x$, e vede se $y$ risulta visitato nel vettore `visited`; la complessità è quella del `dfs`, ovvero $O(n + m)$

```rust
fn does_path_exist(graph: &[Vec<usize>], x: usize, y: usize) -> bool {
    let mut visited = vec![false; graph.len()];
    dfs(graph, x, &mut visited);

    visited[y]
}

fn dfs(graph: &[Vec<usize>], x: usize, visited: &mut Vec<bool>) {
    visited[x] = true;

    for &y in &graph[x] {
        if !visited[y] {
            dfs(graph, y, visited)
        }
    }
}
```
Alternativa **iterativa** del **dfs** che voglio disperatamente semplificare 🫥

```rust
fn dfs_iterative(graph: &[Vec<usize>], x: usize) -> Vec<bool> {
    let mut stack = Vec::from([x]);
    let mut visited = vec![false; graph.len()];
    let mut adjacent = vec![0; graph.len()];

    while let Some(&x) = stack.last() {
        if let Some(&y) = graph[x].get(adjacent[x]) {
            if !visited[y] {
                stack.push(y);
                visited[y] = true;
            }

            adjacent[x] += 1;
        } else {
            stack.pop();
        }
    }

    visited
}
```

**Comment by LeoCosta02** [comment]
```C
main_cammino(Graph: G , x:vert , y: vert  ){
 int[n] vis: [0...0]
  dfs(G,x)
   if(vis[y]==1)
    return true 
    else
     return false
  
}

dfs(G,x){

    vis[x]=1
    for each (y appartenente a x.adiacente){
        if(vis[y]==0)
         dfs(G,y)
    }
}
```

**Reply by IonutCicio** [reply]
> [!TIP]
> Ti suggerisco di provare ad usare VSCode, ha un'opzione FormatOnSave _(basta andare sulle impostazioni)_. Se non erro dovresti scaricare anche l'estensione per C/C++ o il linguaggio che preferisci perché la formattazione funzioni.

![image](https://github.com/sapienzastudentsnetwork/progalg2324/assets/40034429/8c0068d4-ec65-4c33-b788-9f6fea3aa0fa)

> [!TIP]
> Se hai un blocco di codice, puoi fare così per far vedere i colori e farlo tutto carino usando il carattere `` ` `` e specificando il linguaggio, come nell'esempio sotto

````
```C
void blocco_di_codice() {
    int a = 12;
    int b = a + 20;
}
```
````

Non per altro, ma già di suo verificare un algoritmo è complicato, se poi il codice è tutto disordinato mi è difficile dare feedback 🫥

**Reply by LeoCosta02** [reply]
@CuriousCI grazie del consiglio, l’ho editato meglio

**Comment by Warcophyr** [comment]
```python

def dfs(graph, x, vis):
    vis[x] = True

    for y in graph[x]:
        if vis[y] != True:
            dfs(graph, y, vis)

            
def does_path_exits(graph, x, y) -> bool:
    vis = len(graph)*[False]
    dfs(graph, x, vis)

    return vis[y]

```
analogo a @CuriousCI

**Reply by IonutCicio** [reply]
`vis[y] != True` si può accorciare a `!vis[y]` o, ancora meglio, `not vis[y]`; per il resto, è fatto molto bene

**Comment by daromacarlo** [comment]
```c
def esiste_cammino(G, x, y)
	vis = [0,...,0]
	stack = []
	stack.push(x)
	vis[x] = 1
	while stack is not empty{
		v = stack.pop()
		for w in v.adiacenti(){
			if vis[w] == 0{
				vis[x] = 1
				stack.push(w)
			}
			if w == y{
				return true
			}
		}
	}
	return false

//O(n+m)
```
